RoleHierarchy Comments are misleading

<!--
For Security Vulnerabilities, please use https://pivotal.io/security#reporting
-->

### Summary

<!-- 
Please provide a high level summary of the issue you are having
-->
In the current version (5.1.5), the comments in the `org.springframework.security.access.hierarchicalroles.RoleHierarchy` class do not indicate their true behavior and are misleading.

### Actual Behavior

<!-- 
Please describe step by step the behavior you are observing
-->
The correct expression for "role inheritance" should be defined like this: "ROLE_HIGHEST > ROLE_HIGHER > ROLE_LOW > ROLE_LOWER > ROLE_LOWEST".

### Expected Behavior

<!--
Please describe step by step the behavior you expect
-->
But the comment says it needs to be defined like this: "ROLE_HIGHEST > ROLE_HIGHER and ROLE_HIGHER > ROLE_LOW and ROLE_LOW > ROLE_LOWER and ROLE_LOWER > ROLE_LOWEST".  This way of defining expressions is fine in earlier versions, but it is wrong in the current version (I learned from debugging source code).

### Configuration

<!--
Please provide any configuration you have.
-->

### Version
5.1.5
<!--
Please describe what version you are using. Does the problem occur in other versions?
-->

### Sample
Please focus on two places:
`org.springframework.security.access.hierarchicalroles.RoleHierarchyImpl#buildRolesReachableInOneStepMap`
`org.springframework.security.access.expression.SecurityExpressionRoot#hasAnyAuthorityName`
<!--
Providing a complete sample (i.e. link to a github repository) will give this issue higher
priority than issues that do not have a complete sample
-->

> Can you you think the issue is? I have added tests to demonstrate this works as expected See [1f7ba47](https://github.com/spring-projects/spring-security/commit/1f7ba47ce9298950c02b88050207112e25b62180) for the tests

Sorry, I may not have stated it clearly. I mean, in the comments on the "RoleHierarchy" interface, the expression "role inheritance" needs to have the "and" connector to connect, but in fact, it is not correct to use the "and" connection. (or not exactly right, it was correct in earlier versions)

```
public interface RoleHierarchy {

	/**
	 * Returns an array of all reachable authorities.
	 * <p>
	 * Reachable authorities are the directly assigned authorities plus all authorities
	 * that are (transitively) reachable from them in the role hierarchy.
	 * <p>
	 * Example:<br>
	 * Role hierarchy: ROLE_A &gt; ROLE_B and ROLE_B &gt; ROLE_C.<br>
	 * Directly assigned authority: ROLE_A.<br>
	 * Reachable authorities: ROLE_A, ROLE_B, ROLE_C.
	 *
	 * @param authorities - List of the directly assigned authorities.
	 * @return List of all reachable authorities given the assigned authorities.
	 */
	public Collection<? extends GrantedAuthority> getReachableGrantedAuthorities(
			Collection<? extends GrantedAuthority> authorities);

}
```

Please pay attention to the "and" in the comment above.

```
...
Role hierarchy: ROLE_A &gt; ROLE_B and ROLE_B &gt; ROLE_C.<br>
...
```

Unit testing also shows that it is wrong to connect with "and"：

```
	@Test
	public void testRoleHierarchyInterfaceJavadoc() {
		List<GrantedAuthority> flatAuthorities = AuthorityUtils.createAuthorityList("ROLE_HIGHEST");
		List<GrantedAuthority> allAuthorities = AuthorityUtils.createAuthorityList("ROLE_HIGHEST", "ROLE_HIGHER", "ROLE_LOW", "ROLE_LOWER");
		RoleHierarchyImpl roleHierarchyImpl = new RoleHierarchyImpl();
		roleHierarchyImpl.setHierarchy("ROLE_HIGHEST > ROLE_HIGHER and ROLE_HIGHER > ROLE_LOW and ROLE_LOW > ROLE_LOWER");

		assertThat(roleHierarchyImpl.getReachableGrantedAuthorities(flatAuthorities)).containsExactlyInAnyOrderElementsOf(allAuthorities);
	}
```

```
java.lang.AssertionError: 
Expecting:
  <[ROLE_HIGHER and ROLE_HIGHER, ROLE_LOWER, ROLE_LOW and ROLE_LOW, ROLE_HIGHEST]>
to contain exactly in any order:
  <[ROLE_HIGHEST, ROLE_HIGHER, ROLE_LOW, ROLE_LOWER]>
elements not found:
  <[ROLE_HIGHER, ROLE_LOW]>
and elements not expected:
  <[ROLE_HIGHER and ROLE_HIGHER, ROLE_LOW and ROLE_LOW]>


	at org.springframework.security.access.hierarchicalroles.RoleHierarchyImplTests.testRoleHierarchyInterfaceJavadoc(RoleHierarchyImplTests.java:269)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)


Process finished with exit code -1

```

--- a/core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchy.java
+++ b/core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchy.java
@@ -33,7 +33,7 @@ public interface RoleHierarchy {
 	 * that are (transitively) reachable from them in the role hierarchy.
 	 * <p>
 	 * Example:<br>
-	 * Role hierarchy: ROLE_A &gt; ROLE_B and ROLE_B &gt; ROLE_C.<br>
+	 * Role hierarchy: ROLE_A &gt; ROLE_B &gt; ROLE_C.<br>
 	 * Directly assigned authority: ROLE_A.<br>
 	 * Reachable authorities: ROLE_A, ROLE_B, ROLE_C.
 	 *

